src/expr/macro/src/gen.rs

```diff
@@ -423,9 +425,188 @@ impl FunctionAttr {
+        let child: Vec<_> = ids.iter().map(|i| format_ident!("child{i}")).collect();
+        let array_refs: Vec<_> = ids.iter().map(|i| format_ident!("array{i}")).collect();
+        let arrays: Vec<_> = ids.iter().map(|i| format_ident!("a{i}")).collect();
+        let arg_arrays = self
+            .args
+            .iter()
+            .map(|t| format_ident!("{}", types::array_type(t)));
+        let array_builder = format_ident!("{}Builder", types::array_type(&self.ret));
+        let const_arg = match &self.prebuild {
+            Some(_) => quote! { &self.const_arg },
+            None => quote! {},
+        };
+        let const_arg_type = match &self.prebuild {
+            Some(s) => s.split("::").next().unwrap().parse().unwrap(),
+            None => quote! { () },
+        };
+        let const_arg_value = match &self.prebuild {
+            Some(s) => s
+                .replace('$', "child")
+                .parse()
+                .expect("invalid prebuild syntax"),
+            None => quote! { () },
+        };
+        let value = match self.user_fn.return_type {
+            ReturnType::T => quote! { Some(value) },
+            ReturnType::Option => quote! { value },
+            ReturnType::Result => quote! { Some(value?) },
+            ReturnType::ResultOption => quote! { value? },
+        };
+
+        Ok(quote! {
+            |return_type, chunk_size, children| {
+                use risingwave_common::array::*;
+                use risingwave_common::types::*;
+                use risingwave_common::buffer::Bitmap;
+                use risingwave_common::util::iter_util::ZipEqFast;
+                use itertools::multizip;
+
+                crate::ensure!(children.len() == #num_args);
+                let mut iter = children.into_iter();
+                #(let #all_child = iter.next().unwrap();)*
+                #(let #const_child = #const_child.eval_const()?;)*
+
+                #[derive(Debug)]
+                struct #struct_name {
+                    return_type: DataType,
+                    chunk_size: usize,
+                    #(#child: BoxedExpression,)*
+                    const_arg: #const_arg_type,
+                }
+                #[async_trait::async_trait]
+                impl crate::table_function::TableFunction for #struct_name {
+                    fn return_type(&self) -> DataType {
+                        self.return_type.clone()
+                    }
+                    async fn eval<'a>(&'a self, input: &'a DataChunk) -> BoxStream<'a, Result<DataChunk>> {
+                        self.eval_inner(input)
+                    }
+                }
+                impl #struct_name {
+                    #[try_stream(boxed, ok = DataChunk, error = ExprError)]
+                    async fn eval_inner<'a>(&'a self, input: &'a DataChunk) {
+                        #(
+                        let #array_refs = self.#child.eval_checked(input).await?;
```

Comment:
Yes. But I'm afraid it would increase the code complexity. Anyway I'll have a try. 🥵

> BTW, I think calling eval_checked is easy to be missed. Maybe we should use an expression wrapper for doing this.

Could you give an example to clarify it?

Reply:
Something like the wrapper executor. 🥵

https://github.com/risingwavelabs/risingwave/blob/f71ff76e7a471325c8185efbc7005f9fbfa962bb/src/stream/src/task/stream_manager.rs#L548-L557